core/src/main/scala/kafka/server/KafkaConfig.scala

```
@@ -513,6 +513,11 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
     require(replicaFetchWaitMaxMs <= replicaLagTimeMaxMs, "replica.fetch.wait.max.ms should always be less than or equal to replica.lag.time.max.ms" +
       " to prevent frequent changes in ISR")
 
+    if (brokerHeartbeatIntervalMs * 2 > brokerSessionTimeoutMs) {
+      error(s"${KRaftConfigs.BROKER_HEARTBEAT_INTERVAL_MS_CONFIG} ($brokerHeartbeatIntervalMs ms) must be less than or equal to half of the ${KRaftConfigs.BROKER_SESSION_TIMEOUT_MS_CONFIG} ($brokerSessionTimeoutMs ms). " +
```

The message says "must be less than or equal to half", but the server does not actually stop. Perhaps the description could be phrased more clearly?

---

@chia7712 do you think that it's better to throw error if `broker.heartbeat.interval.ms` is larger than half of `broker.session.timeout.ms`? Or we can change the error message like following:

```
      error(s"${KRaftConfigs.BROKER_HEARTBEAT_INTERVAL_MS_CONFIG} ($brokerHeartbeatIntervalMs ms) is larger than half of the ${KRaftConfigs.BROKER_SESSION_TIMEOUT_MS_CONFIG} ($brokerSessionTimeoutMs ms). " +
        "If missing anyone heartbeat request, the broker loses broker lease. " +
        s"Please increase ${KRaftConfigs.BROKER_SESSION_TIMEOUT_MS_CONFIG} or decrease ${KRaftConfigs.BROKER_HEARTBEAT_INTERVAL_MS_CONFIG}.")
```

---

Yes, it would be better to fail fast. However, since session timeout is used by the quorum controller, the broker itself is not aware of the actual session timeout configured for the quorum controller. Perhaps we could improve the error message to make this distinction cleaer, incorporating the above explanation

---

@FrankYang0529 any updates? it would be useful to remind users that session timeout is controlled by quorum controller rather than broker

---

Sorry for late. Update it.